rs-matter/src/transport/exchange.rs

```diff
@@ -1,432 +1,1038 @@
+    /// message is not cleared yet using `Exchange::rx_done` or via some of the `Exchange::send*` methods,
+    /// the method will return an error.
+    ///
+    /// Note that if the uderlying session or exchange tracked by the Matter stack is dropped
+    /// (say, because of lack of resources or a hard networking error), the method will return an error.
+    #[inline(always)]
+    pub async fn recv(&mut self) -> Result<RxMessage<'_>, Error> {
+        if self.rx.is_none() {
+            self.id.recv(self.matter).await
+        } else {
+            Err(ErrorCode::InvalidState.into())
+        }
+    }
 
-            ctx.state = ExchangeState::ExchangeSend {
-                tx: tx as *const _,
-                rx: rx as *mut _,
-                notification: &_self.notification as *const _,
-            };
-            _self.matter.send_notification.signal(());
+    /// Get access to the pending RX message on this exchange, and consume it
+    /// by copying the payload into the provided `WriteBuf` instance.
+    ///
+    /// A syntax sugar for calling ```self.recv().await?``` and then copying the payload.
+    ///
+    /// Returns the exchange message meta-data.
+    ///
+    /// If there is no pending RX message, the method will wait indefinitely until one appears.
+    ///
+    /// If there is already a pending RX message, which was also fetched using `Exchange::recv_fetch` and this
+    /// message is not cleared yet using `Exchange::rx_done` or via some of the `Exchange::send*` methods,
+    /// the method will return an error.
+    ///
+    /// Note that if the uderlying session or exchange tracked by the Matter stack is dropped
+    /// (say, because of lack of resources or a hard networking error), the method will return an error.
+    #[inline(always)]
+    pub async fn recv_into(&mut self, wb: &mut WriteBuf<'_>) -> Result<MessageMeta, Error> {
+        let rx = self.recv().await?;
+
+        wb.reset();
+        wb.append(rx.payload())?;
+
+        Ok(rx.meta())
+    }
 
+    /// Get access to the pending RX message on this exchange, and store it inside the `Exchange` instance, so that
+    /// it is available immediately to consumers by using `Exchange::rx`.
+    ///
+    /// If there is no pending RX message, the method will wait indefinitely until one appears.
+    ///
+    /// If there is already a pending RX message, which was fetched using `Exchange::recv_fetch` and this
+    /// message is not cleared yet using `Exchange::rx_done` or via some of the `Exchange::send*` methods,
+    /// the method will return an error.
+    ///
+    /// Note that if the uderlying session or exchange tracked by the Matter stack is dropped
+    /// (say, because of lack of resources or a hard networking error), the method will return an error.
+    #[inline(always)]
+    pub async fn recv_fetch(&mut self) -> Result<(), Error> {
+        if self.rx.is_none() {
```

Comment:
rust style question: as I review this I find it somewhat harder to review if/else with returns compared to match statements or shorter code.

I would think something like:

```rust
self.rx = match self.rx {
    None => Some(self.id.recv(self.matter).await?),
    Some(_) => Err(ErrorCode::InvalidState)?,
};
Ok(())
```

What are your thoughts on this (i.e. using matching instead of is_some/is_none)? 

To be clear: I believe existing code is correct and I can clearly read and understand what it does. At the same time there seem to be many ways to express the same thing and it seems to be I should prefer chaining (map/ok_or_else/unwrap_or_else ...) if applicable and if not matching on patterns seems to lead to shorter code (and I do not need to look up when seeing else to understand what the else means). Looking for feedback here.

Reply:
The semantics of `recv_fetch` is changed now to return the _previous_ RX packet, if there is one already fetched (rather than returning an error). So I'm not sure a `match` statement is contributing extra clarity anymore.